From 8d3d5b6a9ad934df5a50c3dd0e8e759f331bcd69 Mon Sep 17 00:00:00 2001
From: Haibo Chen <haibo.chen@nxp.com>
Date: Tue, 26 Apr 2016 11:21:02 +0800
Subject: [PATCH 1032/5242] MLK-12706-1 mmc: sdio: add sdio reset function for
 bcmdhd wifi

commit  feeb7acfcdcf5c0aef2d6644be2d44808eb31ee9 from
https://source.codeaurora.org/external/imx/linux-imx.git

This patch add function sdio_reset_comm() to support bcmdhd wifi
dirver build-in type.

Signed-off-by: Haibo Chen <haibo.chen@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/mmc/core/sdio.c |   31 +++++++++++++++++++++++++++++++
 1 file changed, 31 insertions(+)

diff --git a/drivers/mmc/core/sdio.c b/drivers/mmc/core/sdio.c
index bceff55..ec4f4f4 100644
--- a/drivers/mmc/core/sdio.c
+++ b/drivers/mmc/core/sdio.c
@@ -829,6 +829,37 @@ static int mmc_sdio_reinit_card(struct mmc_host *host, bool powered_resume)
 				  powered_resume);
 }
 
+int sdio_reset_comm(struct mmc_card *card)
+{
+	struct mmc_host *host = card->host;
+	u32 ocr;
+	u32 rocr;
+	int err;
+
+	mmc_claim_host(host);
+	mmc_go_idle(host);
+	mmc_set_clock(host, host->f_min);
+	err = mmc_send_io_op_cond(host, 0, &ocr);
+	if (err)
+		goto err;
+	rocr = mmc_select_voltage(host, ocr);
+	if (!rocr) {
+		err = -EINVAL;
+		goto err;
+	}
+	err = mmc_sdio_init_card(host, rocr, card, 0);
+	if (err)
+		goto err;
+	mmc_release_host(host);
+	return 0;
+err:
+	pr_err("%s: Error resetting SDIO communications (%d)\n",
+		mmc_hostname(host), err);
+	mmc_release_host(host);
+	return err;
+}
+EXPORT_SYMBOL(sdio_reset_comm);
+
 /*
  * Host is being removed. Free up the current card.
  */
-- 
1.7.9.5

